src/taipy/gui_core/GuiCoreLib.py

```diff
@@ -11,50 +11,109 @@
+    __PROP_SCENARIO_DATE = "date"
+    __PROP_SCENARIO_NAME = "name"
+    __SCENARIO_PROPS = (__PROP_SCENARIO_CONFIG_ID, __PROP_SCENARIO_DATE, __PROP_SCENARIO_NAME)
     _CORE_CHANGED_NAME = "core_changed"
-    _VAR_NAME = "__CCCtx"
+    _ERROR_VAR = "gui_core_error"
 
-    def __init__(self, gui: Gui, core: Core) -> None:
+    def __init__(self, gui: Gui) -> None:
         self.gui = gui
-        self.core = core
-        self.scenarios: t.List[Scenario]
+        self.scenarios: t.Optional[
+            t.List[t.Tuple[str, str, int, bool, t.Optional[t.List[t.Tuple[str, str, int, bool, None]]]]]
+        ] = None
+        self.scenario_configs: t.Optional[t.List[t.Tuple[str, str]]] = None
 
     def get_scenarios(self):
-        return []
+        if self.scenarios is None:
+
+            def add_scenarios(res, scens):
+                for scenario in scens:
+                    res.append((scenario.id, scenario.name, 1, scenario.is_primary, None))
+                return res
+
+            self.scenarios = []
+            for cycle, scenarios in tp.get_cycles_scenarios().items():
+                if cycle is None:
+                    add_scenarios(self.scenarios, scenarios)
+                else:
+                    self.scenarios.append((cycle.id, cycle.name, 0, False, add_scenarios([], scenarios)))
+        return self.scenarios
+
+    def get_scenario_configs(self):
+        if self.scenario_configs is None:
+            configs = tp.Config.scenarios
+            if isinstance(configs, dict):
+                self.scenario_configs = [(id, f"{c.id}") for id, c in configs.items()]
+        return self.scenario_configs
 
     def create_new_scenario(self, state: State, id: str, action: str, payload: t.Dict[str, str]):
-        pass
+        print(f"create_new_scenario(state, {id}, {action}, {payload}")
+        args = payload.get("args")
+        if args is None or not isinstance(args, list) or len(args) == 0 or not isinstance(args[0], dict):
+            return
+        data = args[0]
+        config_id = data.get(GuiCoreContext.__PROP_SCENARIO_CONFIG_ID)
+        scenario_config = tp.Config.scenarios.get(config_id)
+        if scenario_config is None:
+            state.assign(GuiCoreContext._ERROR_VAR, f"Invalid configuration id ({config_id})")
+            return
+        date_str = data.get(GuiCoreContext.__PROP_SCENARIO_DATE)
+        try:
+            date = parser.parse(date_str) if isinstance(date_str, str) else None
+        except Exception as e:
+            state.assign(GuiCoreContext._ERROR_VAR, f"Invalid date ({date_str}).{e}")
+            return
+        scenario = tp.create_scenario(scenario_config, date, data.get(GuiCoreContext.__PROP_SCENARIO_NAME))
+        if props := data.get("properties"):
+            for prop in props:
+                key = prop.get("key")
+                if key and key not in GuiCoreContext.__SCENARIO_PROPS:
+                    scenario._properties[key] = prop.get("value")
```

Comment:
That s an optimization. When updating an attribute or property of the scenario, the scenario is automatically saved. Using the `with scenario` context, the save is only performed at the exit.

```suggestion
        if props := data.get("properties"):
            with scenario as sc:
                for prop in props:
                    key = prop.get("key")
                    if key and key not in GuiCoreContext.__SCENARIO_PROPS:
                        sc._properties[key] = prop.get("value")
```

Reply:
tx